daft/runners/ray_tracing.py

```diff
@@ -0,0 +1,654 @@
+
+        self._write_event(
+            {
+                "name": "next(tasks)",
+                "pid": 1,
+                "tid": 1,
+                "ph": PHASE_DURATION_END,
+                "args": args,
+            }
+        )
+
+    def task_created(self, task_id: str, stage_id: int, resource_request: ResourceRequest, instructions: str):
+        created_ts = self._write_event(
+            {
+                "id": task_id,
+                "category": "task",
+                "name": f"task_execution.stage-{stage_id}",
+                "ph": PHASE_ASYNC_BEGIN,
+                "args": {
+                    "task_id": task_id,
+                    "resource_request": {
+                        "num_cpus": resource_request.num_cpus,
+                        "num_gpus": resource_request.num_gpus,
+                        "memory_bytes": resource_request.memory_bytes,
+                    },
+                    "stage_id": stage_id,
+                    "instructions": instructions,
+                },
+                "pid": 1,
+                "tid": 1,
+            }
+        )
+
+        if stage_id not in self._stage_start_end:
+            self._stage_start_end[stage_id] = (created_ts, created_ts)
+
+    def task_dispatched(self, task_id: str):
+        self._write_event(
+            {
+                "id": task_id,
+                "category": "task",
+                "name": "task_dispatch",
+                "ph": PHASE_ASYNC_BEGIN,
+                "pid": 1,
+                "tid": 1,
+            }
+        )
+
+    def task_received_as_ready(self, task_id: str, stage_id: int):
+        self._write_event(
+            {
+                "id": task_id,
+                "category": "task",
+                "name": "task_dispatch",
+                "ph": PHASE_ASYNC_END,
+                "pid": 1,
+                "tid": 1,
+            }
+        )
+        new_end = self._write_event(
+            {
+                "id": task_id,
+                "category": "task",
+                "name": f"task_execution.stage-{stage_id}",
+                "ph": PHASE_ASYNC_END,
+                "pid": 1,
+                "tid": 1,
+            }
+        )
+
+        assert stage_id in self._stage_start_end
+        old_start, _ = self._stage_start_end[stage_id]
+        self._stage_start_end[stage_id] = (old_start, new_end)
+
+
+@dataclasses.dataclass(frozen=True)
+class RayFunctionWrapper:
+    """Wrapper around a Ray remote function that allows us to intercept calls and record the call for a given task ID"""
+
+    f: ray.remote_function.RemoteFunction
+
+    def with_tracing(self, runner_tracer: RunnerTracer, task: PartitionTask) -> RayRunnableFunctionWrapper:
+        return RayRunnableFunctionWrapper(f=self.f, runner_tracer=runner_tracer, task=task)
+
+    def options(self, *args, **kwargs) -> RayFunctionWrapper:
+        return dataclasses.replace(self, f=self.f.options(*args, **kwargs))
+
+    @classmethod
```

Comment:
instead of having wrap be the user abstraction, the pattern should look more like:

```
@dataclasses.dataclass(frozen=True)
class RayFunctionWrapper:
...

def ray_remote_trace(f: ray.remote_function.RemoteFunction):
  return RayFunctionWrapper(f)


```